Exercises/NoModules/Chapter 17/Soln17_05/LinkedList.h

```diff
@@ -0,0 +1,278 @@
+	LinkedList& operator=(LinkedList list);      // Assignment operator
+
+	void push_front(const T& value);             // Add an object to the head
+	void push_back(const T& value);              // Add an object to the tail
+
+	void pop_back();                             // Removes the last element from the list (undefined behavior for empty lists)
+	void pop_front();                            // Removes the first element from the list (undefined behavior for empty lists)
+
+	T& front();                                  // Get the object at the head (undefined behavior for empty lists)
+	T& back();                                   // Get the object at the tail (undefined behavior for empty lists)
+	const T& front() const;                      // Get the object at the head (undefined behavior for empty lists)
+	const T& back() const;                       // Get the object at the tail (undefined behavior for empty lists)
+
+	bool empty() const;                          // Checks whether the list is empty or not
+	void clear();                                // Function to remove all elements from the list
+	size_t size() const;                         // Get the number of elements from the list
+
+	class Iterator;                              // Nested Iterator class declaration (definition below)
+	Iterator front_iterator() const;             // Get an Iterator that starts at the head
+	Iterator back_iterator() const;              // Get an Iterator that starts at the tail
+
+	void swap(LinkedList& other) noexcept;       // Swap function
+
+private:
+	class Node                                   // Node class definition
+	{
+	public:
+		Node(const T& theValue)
+			: m_value{ theValue }, m_next{}, m_previous{}
+		{}
+
+		T m_value;
+		Node* m_next;
+		Node* m_previous;
+	};
+
+	Node* m_head{};    // Pointer to first node
+	Node* m_tail{};    // Pointer to last node
+	size_t m_size{};
+};
+
+// Non-member swap function.
+// Convention dictates swap is present as non-member function.
+// For class types it is often easiest (and recommended) 
+// to implement it using a member swap function.
+template<typename T>
+void swap(LinkedList<T>& one, LinkedList<T>& other)
+{
+	one.swap(other);
+}
+
+// ---------------------------------------------
+// Member definitions
+// ---------------------------------------------
+
+// Destructor template
+template<typename T>
+LinkedList<T>::~LinkedList()
+{
+	clear();
+}
+
+// Copy constructor template
+template<typename T>
+LinkedList<T>::LinkedList(const LinkedList& list)
+{
+  // Use existing members (iteration, push_back()) to implement the copying
+  // This avoids duplicating any code that manipulates the list's pointers / size members.
+  for (auto iterator = list.front_iterator(); iterator; iterator.next())
+    push_back(iterator.value());
+}
+
+// Assignment operator template (uses copy-and-swap idiom)
+template<typename T>
+LinkedList<T>& LinkedList<T>::operator=(LinkedList copy)
```

Comment:
usually, you pass by const reference and make a copy in the body.
Why now different?

Reply:
Living on the edge I suppose.